remote_server/remote_server.rb

```
@@ -0,0 +1,409 @@
+    CopperEgg::Api.host = @config['copperegg']['apihost'] unless @config['copperegg']['apihost'].nil?
+    @freq = @config['copperegg']['frequency'] unless @config['copperegg']['frequency'].nil?
+    @services = @config['copperegg']['services']
+  else
+    log 'You have no copperegg entry in your config.yml!'
+    log 'Edit your config.yml and restart.'
+    exit
+  end
+end
+
+if CopperEgg::Api.apikey.nil?
+  log 'You need to supply an apikey with the -k option or in the config.yml.'
+  exit
+end
+
+if @services.empty?
+  log 'No services listed in the config file.'
+  log 'Nothing will be monitored!'
+  exit
+end
+
+@freq = 60 unless [60, 300, 900, 3600, 216_00].include?(@freq)
+
+log "Update frequency set to #{@freq}s."
+
+
+trap('INT') { parent_interrupt }
+trap('TERM') { parent_interrupt }
+
+MAX_RETRIES = 30
+last_failure = 0
+
+MAX_SETUP_RETRIES = 5
+setup_retries = MAX_SETUP_RETRIES
+
+begin
+  dashboards = CopperEgg::CustomDashboard.find
+  metric_groups = CopperEgg::MetricGroup.find
+rescue => e
+  log "Error connecting to server.  Retrying (#{retries}) more times..."
+  raise e if @debug
+
+  sleep 2
+  setup_retries -= 1
+  retry if setup_retries > 0
+  # If we can't succeed with setup on the servcies, let's just error out
+  raise e
+end
+
+@services.each do |service|
+  if @config[service] && !@config[service]['servers'].empty?
+    begin
+      log "Checking for existence of metric group for #{service}"
+      metric_group = metric_groups.nil? ? nil : metric_groups.detect {|m| m.name == @config[service]['group_name']}
+      metric_group = ensure_metric_group(metric_group, service)
+      raise "Could not create a metric group for #{service}" if metric_group.nil?
+
+      log "Checking for existence of #{service} Dashboard"
+      dashboard = dashboards.nil? ? nil : dashboards.detect { |d| d.name == @config[service]['dashboard'] } || create_dashboard(service, metric_group)
+      log "Could not create a dashboard for #{service}" if dashboard.nil?
```

raise instead of log